syntax = "proto3";
package DDRCommProto;

// version 10001, 2019/05/11

enum eRspCode
{
	eOkay = 0;
	eReqVoid = 1; // requests invalid themselves
	eTooFrequent = 2; // route to upload is too close to other routes from the same robot
	eIDInvalid = 3; // invalid upload/download ID
	eDecompFailed = 4; // failed to decompress contents
	eFileNotExisting = 5;
	eFilePositionIncorrect = 6; // request of incorrect file positions	
}

enum eZipType
{
	eNoZip = 0;
	eZLib = 1;
}

message OneRouteInfo
{
	int32 version = 1;
	string robotID = 2;
	string routeName = 3;
	int64 dateTime = 4; // route creation time (seconds since EPOCH)
}

message reqCreateRoute
{
	OneRouteInfo routeInfo = 1;
}

message rspCreateRoute
{
	eRspCode ret = 1;
	int64 uploadID = 2;
}

// requests to upload a single file or multiple files
message reqUploadFiles
{
	int64 reqGUID = 1;
	int64 uploadID = 2;
	repeated string fileNames = 3;
	repeated int64 fileSz = 4; // list of file sizes (in number of bytes)
	repeated int64 fileMTimes = 5; // list of file modification times (in seconds since EPOCH)
	int64 file1_startPos = 6; // starting position of the first file (0 as default as the beginning of this file)
	int64 fileN_endPos = 7; // ending position of the last file (-1 default as the end of this file)
	eZipType zipType = 8;
	int64 content_len = 9; // length of original contents
	bytes contents = 10; // contents of files after decompression
}

message rspUploadFiles
{
	eRspCode ret = 1;
	int64 reqGUID = 2;
}

message reqListRoutes
{
}

message rspListRoutes
{
	eRspCode ret = 1;
	repeated OneRouteInfo routes = 2;
	repeated int64 downloadID = 3;
}

message reqListFiles_uploader
{
	int64 uploadID = 1;
}

message reqListFiles_downloader
{
	int64 downloadID = 1;
}

message rspListFiles
{
	eRspCode ret = 1;
	repeated string fileNames = 2;
	repeated int64 fileSz = 3;
	repeated int64 fileCurPos = 4; // may be < fileSz
	repeated int64 fileMTimes = 5; // list of file modification times (in seconds since EPOCH)
}

message reqDownloadFiles
{
	int64 reqGUID = 1;
	int64 downloadID = 2;
	repeated string fileNames = 3;
	int64 file1_startPos = 4; // starting position of the first file (0 as default as the beginning of this file)
	int64 fileN_endPos = 5; // ending position of the last file (-1 default as the end of this file)
}

message rspDownloadFiles
{
	eRspCode ret = 1;
	int64 reqGUID = 2;
	repeated string fileNames = 3;
	repeated int64 fileSz = 4;
	int64 file1_startPos = 5; // starting position of the first file (0 as default as the beginning of this file)
	int64 fileN_endPos = 6; // ending position of the last file (-1 default as the end of this file)
	eZipType zipType = 7;
	int64 content_len = 8; // length of original contents
	bytes contents = 9; // contents of files after decompression
}

message hbFrame
{
}

